src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/seastar.hh>
+
+#include <bits/types/FILE.h>
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+
+namespace cloud_storage {
+
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period)
+  : _cache_dir(std::move(cache_dir))
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period) {}
+
+ss::future<> cache::start() {
+    // todo: start maintenance cycle here, until then don't delete this print
+    // statements so that compiler is not complaining about unused fields
+    std::cout << "max_cache_size: " << _max_cache_size << "\n";
+    std::cout << "check_period: " << _check_period.count() << "\n";
+    return ss::make_ready_future<>();
+}
+
+ss::future<> cache::stop() { return _gate.close(); }
+
+ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    gate_guard guard{_gate};
+    ss::file cache_file;
+    try {
+        cache_file = co_await ss::open_file_dma(
+          (_cache_dir / key).native(), ss::open_flags::ro);
+    } catch (...) {
+        vlog(
+          cst_log.debug,
+          "Couldn't open file {} to read: {}",
+          (_cache_dir / key).native(),
+          std::current_exception());
+        co_return std::nullopt;
+    }
+
+    auto data_size = co_await cache_file.size();
+    auto data_stream = ss::make_file_input_stream(cache_file);
+    co_return std::optional(cache_item{std::move(data_stream), data_size});
+}
+
+ss::future<>
+cache::put(std::filesystem::path key, ss::input_stream<char> data) {
+    gate_guard guard{_gate};
+
+    auto filename = (_cache_dir / key).filename();
+    auto tmp_filename = std::filesystem::path(filename.string() + ".part");
+    auto dir_path = (_cache_dir / key).remove_filename();
+
+    co_await ss::recursive_touch_directory(dir_path.string());
+
+    auto flags = ss::open_flags::wo | ss::open_flags::create;
```

Comment:
what would happen if the system crashed after creating the file, but before renaming the file?

```
co_await ss::open_file_dma((dir_path / tmp_filename).native(), flags);
... <crash> ...
co_await ss::rename_file((dir_path / tmp_filename).native(), (dir_path / filename).native());
```

the file would be on disk, but if a `key` were reused would the next cache::put _append_ data to a previous instance of the cached data instead of writing into an empty file (which is what it looks like the intention is). you could use O_EXCL to make it a bit more robust, or combine O_EXCL with some extra protection from adding in a random value to the file name.

even if the `key` is intended to solve this issue, it is probably good to use `O_EXCL` so there is some runtime checking of assumptions.

Reply:
**tl;dr:** Tried to reproduce behavior when the data is _appended_, but didn't succeed.

If the system is crashed after creating file, but before renaming it, we will have incomplete `key.part` file that no one is using. Next call to `put(key)` will open `key.part` with `ss::open_flags::wo` will rewrite the `key.part` file, then rename it to `key`, so will have the latest data in key file.

Reply:
`ss::file_output_stream` will probably start from the beginning which means that if the old value is longer than the new one, there is a potential for the file containing both the new and (the suffix of) the old value. So O_TRUNC is probably needed.

Reply:
added O_TRUNC flag